src/librustc/infer/error_reporting.rs

```diff
@@ -560,6 +560,27 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                     diag.note_expected_found(&"type", &expected, &found);
                 }
             }
+
+            if let Some((found, (expected_ty, _))) = self.get_ids(values) {
+                // look for expected with found id
+                self.tcx.populate_inherent_implementations_for_type_if_necessary(found);
+                if let Some(impl_infos) = self.tcx.inherent_impls.borrow().get(&found) {
+                    let mut methods = Vec::new();
+                    for impl_ in impl_infos {
+                        methods.append(&mut self.tcx
```

Comment:
I am sure that this can be replaced with flat_map, isn't it? This will reduce allocations, I think.